Add methods for dealing with fairness in rankings

andrewklayk opened the pull request:
We've implemented a set of post-processing algorithms for dealing with fairness in rankings, as described in https://arxiv.org/abs/1905.01989 (KDD 2019), complete with a demo notebook and tests.
Along with the algorithms, we've added two useful metrics for evaluating rankings that were introduced or referenced in the paper (Infeasibility Index and Discounted Cumulative Gain). Further details are available in the demo notebook.

closes #455

hoffmansc commented on aif360/algorithms/postprocessing/deterministic_reranking.py:
Can you add a line to [`docs/source/modules/algorithms.rst`](https://github.com/Trusted-AI/AIF360/blob/master/docs/source/modules/algorithms.rst) in order to show the API documentation for this?

hoffmansc commented on examples/demo_deterministic_reranking.ipynb:
```diff
@@ -0,0 +1,1103 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To initialize the DeterministicReranking class, we need to pass the **protected attribute values** of the priviliged and unprivileged groups.\n",
```
"privileged" is misspelled a couple places

hoffmansc reviewed (changes requested):
thanks @andrewklayk, this looks great! Other than the small comments, I think we can merge this.

I would love to include this in the `aif360.sklearn` subpackage as well. There is a slightly different API for those algorithms/metrics but it should actually be easier to use since you rely on DataFrames anyway. That can be a separate PR if you prefer or you can add it to this one.

hoffmansc commented on aif360/algorithms/postprocessing/deterministic_reranking.py:
```diff
@@ -0,0 +1,212 @@
+    """
+
+    def __init__(self,
+            unprivileged_groups: list[dict],
+            privileged_groups: list[dict]):
```
This type hint doesn't work in Python 3.8. You can change it to use `List` from `typing`. Type hints are optional for this package, though. We just require the type in the docstring

andrewklayk commented on examples/demo_deterministic_reranking.ipynb:
```diff
@@ -0,0 +1,1103 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tempfile\n",
+    "import requests\n",
+    "import zipfile\n",
+    "import os\n",
+    "import pandas as pd \n",
+    "\n",
+    "with tempfile.TemporaryDirectory() as temp_dir:\n",
+    "    response = requests.get(\"http://www.seaphe.org/databases/LSAC/LSAC_SAS.zip\")\n",
+    "    temp_file_name = os.path.join(temp_dir, \"LSAC_SAS.zip\")\n",
+    "    with open(temp_file_name, \"wb\") as temp_file:\n",
+    "        temp_file.write(response.content)\n",
+    "    with zipfile.ZipFile(temp_file_name, 'r') as zip_ref:\n",
+    "        zip_ref.extractall(temp_dir)\n",
+    "    data = pd.read_sas(os.path.join(temp_dir, \"lsac.sas7bdat\"))\n",
+    "    data = data.assign(gender=(data[\"gender\"] == b\"male\") * 1)\n",
+    "    data['race'] = data['race1']\n",
+    "    data = data[['race', 'gender', 'lsat', 'ugpa', 'zfygpa']]\n"
```
That was the initial idea, but I couldn't get the `tempeh` import to work, despite having installed the `.[LawSchoolGPA]` option of the package. I've tried installing `tempeh` separately, reinstalling AIF360 manually and so on, but to no end. Without the import, neither the `LawSchoolGPADataset` nor the `fetch_lawschool_gpa` work.

andrewklayk replied:
An update on this: it seems like the SEAPHE website is no more (domain expired), which means that neither this code nor the AIF360 functions work.

hoffmansc replied:
that is unfortunate... I've reached out to SEAPHE and they're investigating. Let's give it a week and see if they can get the website back up.

andrewklayk commented:
Hello @hoffmansc
I've addressed the issues you stated, except for the usage of the SEAPHE dataset. The site is still down, and I've been unable to find a free-use dataset from a reputable source that would fit here as an example. 
In my opinion, the "toy example" in the notebook is enough to explain the API, so perhaps we should ditch the dataset part alltogether?

rahulnair23 commented:
Hi @hoffmansc - seems like the Law School dependency has been removed from the demo notebook. Is this good to merge?
